app/domain/authentication/authn_gcp/authenticator.rb

```diff
@@ -3,29 +3,47 @@
 module Authentication
   module AuthnGcp
 
+    # This authenticator validates that the received token matches the
+    # restrictions defined in the policy. But there is another validation step
+    # where we also verify that the token is valid but this is done while it
+    # is decoded (using a third-party).
     Authenticator = CommandClass.new(
       dependencies: {
-        validate_resource_restrictions: ValidateResourceRestrictions.new,
+        validate_resource_restrictions: Authentication::ResourceRestrictions::ValidateResourceRestrictions.new,
+        authentication_request_class:   AuthenticationRequest,
         logger:                         Rails.logger
       },
       inputs:       [:authenticator_input]
     ) do
 
       extend Forwardable
-      def_delegators :@authenticator_input, :authenticator_name, :account, :username, :credentials
+      def_delegators :@authenticator_input, :authenticator_name, :service_id, :account, :username, :credentials
 
       def call
+        create_authentication_request
         validate_resource_restrictions
       end
 
       private
 
+      def create_authentication_request
+        authentication_request
+      end
+
+      def authentication_request
+        @authentication_request ||= @authentication_request_class.new(
+          decoded_token: credentials
+        )
+      end
+
       def validate_resource_restrictions
         @validate_resource_restrictions.call(
           authenticator_name: authenticator_name,
-          account:     account,
-          username:    username,
-          credentials: credentials
+          service_id: service_id,
+          account: account,
+          host_name: username,
+          constraints: Restrictions::CONSTRAINTS,
```

Comment:
why does constraints need to be passed in here and not just initialized in the calling class?

Reply:
because it is different for each authenticator, and is not a dependency for `ValidateResourceRestrictions`.
The `ValidateResourceRestrictions` CommandClass just calls `validate(resource_restrictions)` on it, but it doesn't know (or care) which restrictions are they.

Reply:
If this is the case do we plan on also having a Restrictions module call azure and k8s?

why can't you just have a line at the top of `authenticator.rb` like so:

```
module Authentication
  module AuthnGcp
  CONSTANTS = ....
```